File: sources/puredata_src/data_reader.cpp
/**
 * @file data_reader.cpp
 * @brief Reads HRTF data from a file and writes it to a C++ header file.
 * @note The functions have been roughly taken from earplug's `new` function.
 * This means that they were originally written in C. To avoid errors, file
 * pointers are still used, instead of C++'s iostream library.
 * @author Zucchelli Maurizio
 * @version 0.0.1
 * @date 2014-07-14
 */

#include <string>

#define POINTS 368
#define CHANNELS 2
#define LENGTH 128

using namespace std;

float values[POINTS][CHANNELS][LENGTH];

void read_data(string filename)
{
	int i, j;
	FILE* fp = fopen(filename.c_str(), "r");  

	for (i = 0; i < POINTS; i++) 
	{
		while(fgetc(fp) != 10);

		for (j = 0; j < LENGTH; j++)
			fscanf(fp, "%f %f ", &values[i][0][j], &values[i][1][j]);
	}

	fclose(fp);
}

void write_data(string filename)
{
	int i, j, k;
	FILE* fp = fopen(filename.c_str(), "w");  

	fprintf(fp, "#include \"m_pd.h\"\n\nextern \"C\" t_float hrtf_impulses[368][2][128] =\n{\n");
	for (i = 0; i < POINTS; i++) 
	{
		fprintf(fp, "\t{\n");
		for (j = 0; j < CHANNELS; j++)
		{
			fprintf(fp, "\t\t{\n");
			for (k = 0; k < LENGTH; k++)
			{
				if (k % 8 == 0)
					fprintf(fp, "\t\t\t");
				else
					fprintf(fp, " ");

				fprintf(fp, "%f", values[i][j][k]);

				if (k % 8 == 7)
					fprintf(fp, "\n");
				else
					fprintf(fp, ",");
			}
			fprintf(fp, "\t\t}");

			if (j < CHANNELS-1)
				fprintf(fp, ",");
			fprintf(fp, "\n");
		}
		fprintf(fp, "\t}");

		if (j < POINTS-1)
			fprintf(fp, ",");
		fprintf(fp, "\n");
	}
	fprintf(fp, "};");

	fclose(fp);
}

int main(int argc, char** argv)
{
	if (argc <= 1)
	{
		printf("Usage: data_reader input [output]\n");
		return 1;
	}

	read_data(argv[1]);

	if (argc <= 2)
		write_data("hrtf_data.hpp");
	else
		write_data(argv[2]);

	return 0;
}

File: sources/puredata_src/data_reader.rb
#!/usr/bin/env ruby

# @brief Reads HRTF data from a file and writes it to a C++ header file.
# @author Zucchelli Maurizio
# @version 0.0.2
# @date 2014-07-15

class Array
  def odd_values
    self.select.each_with_index { |_, i| i.odd? }
  end
  def even_values
    self.select.each_with_index { |_, i| i.even? }
  end
end

class DataReader
  attr_reader :values, :coordinates

  def initialize()  
    @values = Array.new
    @coordinates = Array.new
  end  

  def read_data(filename)
    File.open(filename, 'r').each do |line|
      if /(?<elevation>-?[[:digit:]]+)e(?<angle>[[:digit:]]+)a/ =~ line
        @coordinates.push([elevation.to_i, angle.to_i])
      else
        filter = line.scan(/-?[[:digit:]]+\.[[:digit:]]+[[:space:]]/)
        if filter.length > 0
          @values.push(filter.map{ |i| i.to_f })
        end
      end
    end

    self
  end

  def elaborate_data()
    File.open("data_reader.m", 'w') do |file|
      file.write("input = [\n");
      coordinates.each_with_index do |point, i|
        cartesian = [ Math.cos(point[0]) * Math.cos(point[1]),
                      Math.cos(point[0]) * Math.sin(point[1]),
                      Math.sin(point[0]) ]

        file.write("\t%f, %f, %f;\n" % cartesian )
      end
      file.write("];")
    end

    `matlab -nodisplay -nosplash -r "data_triangulation; exit" > /dev/null`

    `rm data_reader.m`

    self
  end

  def write_data(filename)
    File.open(filename, 'w') do |file|

      # Header
      file.write("#ifndef __HRTF_DATA__\n#define __HRTF_DATA__\n\n#include \"m_pd.h\"\n\n")

      # Values
      file.write("extern \"C\" t_float hrtf_impulses[][2][#{values[0].length/2}] =\n{\n")
      values.each_with_index do |point, i|
        file.write("\t{\n\t\t{\n")

        point.even_values.each_with_index do |value, j|
          if j%8 == 0
            file.write("\t\t\t")
          else
            file.write(" ")
          end

          file.write("%f," % value)

          if j%8 == 7
            file.write("\n")
          end
        end

        file.write("\t\t},\n\t\t{\n")

        point.odd_values.each_with_index do |value, j|
          if j%8 == 0
            file.write("\t\t\t")
          else
            file.write(" ")
          end

          file.write("%f," % value)

          if j%8 == 7
            file.write("\n")
          end
        end

        file.write("\t\t}\n\t}")

        if i < values.length-1
          file.write(",");
        end
        file.write("\n");
      end
      file.write("};\n\n")

      # Coordinates
      file.write("extern \"C\" t_int hrtf_coordinates[][2] =\n{\n")
      coordinates.each_with_index do |point, i|
        file.write("\t{ %d, %d }" % point )

        if i < coordinates.length-1
          file.write(",")
        end
        file.write("\n")
      end

      file.write("};\n\n")
    end

    elaborate_data

    File.open(filename, 'a') do |file|
      # Footer
      file.write("#endif  // __HRTF_DATA__");
    end
  end
end

DataReader.new.read_data(ARGV[0]).write_data("hrtf_data.hpp")
